### PR TITLE
Resolve symlinked modules correctly

### PR DESCRIPTION
**Pull request by lhorie**
Some systems (e.g. rush, pnpm) use symlinks to create a recursive dependency graph
instead of relying on the hoisting aspect of the node module resolution algorithm (like npm and yarn do)
in these cases, we want to resolve to the real path of a module, so that the tree structure below
only ever tries to run the `x` module once (rather than once on each module that depends on it)

```
- node_modules
  - a
    - node_modules
      - symlink to x
  - b
    - node_modules
      - symlink to x
```

There's an example of a downstream bug here: https://github.com/lhorie/jest-browser-bug (via https://github.com/facebook/jest/issues/7840)

### CODE REVIEW

**Comment by lhorie** [comment]
@defunctzombie Ping. Is this package still maintained?

File: package.json [hunk 1]
```diff
@@ -8,7 +8,8 @@
     "empty.js"
   ],
   "scripts": {
-    "test": "mocha --reporter list test/*.js"
+    "test": "mocha --reporter list test/*.js",
+    "postinstall": "node scripts/setup-symlinks.js"
```
Comment (defunctzombie) [inline package.json]: End users don't need this link to happen so I'd prefer if the module didn't do it for all users.

File: index.js [hunk 2]
```diff
@@ -216,6 +216,20 @@ function resolve(id, opts, cb) {
     opts = opts || {};
     opts.filename = opts.filename || '';
 
+    var cb = function(err, path, pkg) {
+        fs.stat(path, function(notPath) {
+            if (notPath) callback(err, path, pkg);
+            else {
+                fs.realpath(path, function(notReal, real) {
+                    if (notReal) callback(err, path, pkg);
```
Comment (defunctzombie) [inline index.js]: Please follow the code style already present in the document.

File: index.js [hunk 3]
```diff
@@ -216,6 +216,20 @@ function resolve(id, opts, cb) {
     opts = opts || {};
     opts.filename = opts.filename || '';
 
+    var cb = function(err, path, pkg) {
```
Comment (defunctzombie) [inline index.js]: What is this set of nested functions trying to do?

Comment (lhorie) [reply index.js]: 1) if it's a module name, resolve to it
2) if it's something in the filesystem, then
  2a) if it's a regular file, resolve to it
  2b) if it's a symlink, resolve to its realpath

**Comment by lhorie** [comment]
@defunctzombie I made the code style change and moved the test setup script to the `test` script

**Comment by rtsao** [comment]
@defunctzombie Is there anything else that needs to be done for this to be ready to merge?

I suppose get tests passing in older version of node?

**Comment by defunctzombie** [comment]
It's fair to say that I don't maintain or keep this module updated very much. Which upstream module(s) depend on this enough to have their maintainers consider picking up this module as well?

**Comment by rtsao** [comment]
I see. Jest is currently using this package (see https://github.com/facebook/jest/pull/9511). I suppose the Jest maintainers might be amenable to taking over ownership (or at the very least forking their own version to address their needs).

**Comment by defunctzombie** [comment]
@rtsao I'd be happy to see this PR land if the tests were updated to pass on the versions of node that jest actively supports - I don't care if old node version support is dropped. If you want to prepare another PR building from this one and ping me I can merge it.

I am also happy to have jest maintainers pickup this module or logic if that is easier for them. I do not commit resources to this module.

**Comment by rtsao** [comment]
Thanks @defunctzombie!

@lhorie was kind of enough to grant me push access so I've just made the changes here.

As of [Jest v25, Node 6.x support was removed](https://jestjs.io/blog/2020/01/21/jest-25#bye-node-6). Additionally, [Node 6.x EOL was April 30, 2019](https://github.com/nodejs/Release#end-of-life-releases). So I've updated the travis.yml to test against 8.x, 10.x and 12.x and added an npm lockfile so that tests will be deterministic and won't break in the future.

**Pull request merged**